### DISCUSSION TITLE
Deploy a local repository of Maven artifacts to Nexus

### DISCUSSION THREAD

**Discussion by gastaldi** [jreleaser/jreleaser#1708] [Q&A]
For security reasons, I have artifacts built in a GitHub workflow and then published in another. 

In the first workflow I have the following content generated in `/tmp/repository` executed with my Maven project
```bash
mvn deploy -DaltDeploymentRepository=local::file://tmp/repository
```

Now, for the second workflow (where publishing is done) can JReleaser help me in publishing `/tmp/repository` to Nexus? If so, can you point me to some example that does that?

**Comment by gastaldi** [comment]
I am looking for something like https://github.com/sonatype/nexus-maven-plugins/tree/main/staging/maven-plugin#deploy-staged-repository, perhaps I can invoke the plugin directly but it would be nice if JReleaser could do that

**Comment by aalmiray** [comment]
Yes, JReleaser can do that. Matter of fact, that's the only way to do it with the tool, that is, stage artifacts in a local directory then release. The guide describes how to do it

https://jreleaser.org/guide/latest/examples/maven/maven-central.html

**Reply by gastaldi** [reply]
Awesome, I did try it but I get this error: 
 
```
Error:  == JReleaser ==
Error:  project.java.groupId must not be blank
JReleaser has not been properly configured.
project.java.groupId must not be blank
```

Probably because there is no pom.xml in the directory where `jreleaser` is executed (I am using the GitHub Action `jreleaser/jreleaser-action@v2` to be more precise), just the staged repository.

Any hints?

**Reply by aalmiray** [reply]
Well, not really. There is no need to have a pom.xml at the same location as the jreleaser.yml file.

However, the error states there's some missing information that should go into the project section of the jreleaser.yml file, similar to what you can see here https://github.com/jreleaser/helloworld-java-bin/blob/main/jreleaser.yml

**Reply by gastaldi** [reply]
My use case is a bit different: I have a `quarkiverse-release` repository with a jreleaser.yml file used to push to Maven Central. Given that any project in the same org can invoke this workflow, is it possible to set the project groupId dynamically?

**Reply by gastaldi** [reply]
Also, can you elaborate on why the groupId is needed while deploying a staged local repository to Maven Central?

**Reply by aalmiray** [reply]
> My use case is a bit different: I have a `quarkiverse-release` repository with a jreleaser.yml file used to push to Maven Central. Given that any project in the same org can invoke this workflow, is it possible to set the project groupId dynamically?  

Not at the moment but could be added as a feature. All secret values may be set using env vars or sys props. Only a few non secret values may be set in the same way. `java.groupId` is not one of them as of the latest release.

**Reply by aalmiray** [reply]
> Also, can you elaborate on why the groupId is needed while deploying a staged local repository to Maven Central?

This is due to validation constraints as the target namespace/groupId must exist at MC and your credentials must have access to it. This step must occur before any artifacts may be uploaded. Also, multiple artifacts related to different groupIds may be uploaded as a bundle, as long as all of them share a common groupId ancestor or namespace.

**Reply by aalmiray** [reply]
> My use case is a bit different: I have a `quarkiverse-release` repository with a jreleaser.yml file used to push to Maven Central. Given that any project in the same org can invoke this workflow, is it possible to set the project groupId dynamically?  

In this case either all quarkiverse projects share the same jreleaser.yml file as long as every entry that may need tweaks can be programmatically updated (env vars or sys props). Might be tricky. I'd suggest instead that every project has it's own jreleaser.yml file even if a large percentage of projects end up customizing only a few properties.

**Reply by gastaldi** [reply]
I created https://github.com/jreleaser/jreleaser/issues/1709 asking `java.groupId` to be set using env vars as you suggested. 

> In this case either all quarkiverse projects share the same jreleaser.yml file as long as every entry that may need tweaks can be programmatically updated (env vars or sys props). Might be tricky. I'd suggest instead that every project has it's own jreleaser.yml file even if a large percentage of projects end up customizing only a few properties.

I'm not using jreleaser.yml in the other Quarkiverse projects (yet - there are plans to do that in the future, but not right now), this is the repository I'm working on: https://github.com/quarkiverse-release/quarkiverse-release. The workflow takes a ZIP file (built from another Quarkiverse project) and must deploy to Maven Central.

**Reply by aalmiray** [reply]
Oh, I understand now. I'm afraid you may need additional properties in the `project` section otherwise validation will fail. On which case those properties would need to be parameterizable as well.

**Reply by gastaldi** [reply]
Maybe a separate discussion, but I am seeing errors like the following when pomchecker is executed:
```bash
[FATAL] Non-resolvable parent POM for io.quarkiverse.slack:quarkus-slack-parent:0.0.1: The following artifacts could not be resolved: io.quarkiverse:quarkiverse-parent:pom:16 (absent): Could not find artifact io.quarkiverse:quarkiverse-parent:pom:16 and 'parent.relativePath' points at wrong local POM @ line 5, column 13
```

It looks like I am missing something, but shouldn't it be resolving these artifacts from Maven Central?

**Reply by aalmiray** [reply]
Yes, that's the case. Pomchecker should attempt resolving files from MC first. Did you try invoking the deploy command with `--dry-run` and inspect `out/jreleaser/trace.log`?

If there happens to be a bug in pomchecker then it should filed at https://github.com/kordamp/pomchecker

**Reply by gastaldi** [reply]
Seems to be a bug, I created https://github.com/kordamp/pomchecker/issues/22 and provided a PR fixing it in https://github.com/kordamp/pomchecker/pull/23
